File: ui/src/js/sheet/transform.test.ts
import { describe, it, expect } from 'vitest';
import { transform } from './transform';
import type { Op } from './op';

const setCell = (row: number, col: number): Op => ({ type: 'setCell', sheet: 's1', baseRev: 0, row, col, raw: 'x' });

describe('transform (port of Go Transform)', () => {
  it('cell below insertRows shifts down; above stays', () => {
    const applied: Op = { type: 'insertRows', sheet: 's1', baseRev: 0, index: 2, count: 3 };
    expect(transform(setCell(4, 0), applied).row).toBe(7);
    expect(transform(setCell(1, 0), applied).row).toBe(1);
  });

  it('cell below deleteRows shifts up; inside band clamps', () => {
    const applied: Op = { type: 'deleteRows', sheet: 's1', baseRev: 0, index: 2, count: 2 };
    expect(transform(setCell(5, 0), applied).row).toBe(3);
    expect(transform(setCell(3, 0), applied).row).toBe(2);
  });

  it('different sheet is a no-op', () => {
    const applied: Op = { type: 'insertRows', sheet: 'other', baseRev: 0, index: 0, count: 5 };
    const out = transform(setCell(1, 1), applied);
    expect(out.row).toBe(1);
    expect(out.col).toBe(1);
  });

  it('insert index shifts against a prior insert', () => {
    const applied: Op = { type: 'insertRows', sheet: 's1', baseRev: 0, index: 2, count: 2 };
    const inOp: Op = { type: 'insertRows', sheet: 's1', baseRev: 0, index: 4, count: 1 };
    expect(transform(inOp, applied).index).toBe(6);
  });

  it('cell right of insertCols shifts right', () => {
    const applied: Op = { type: 'insertCols', sheet: 's1', baseRev: 0, index: 1, count: 2 };
    expect(transform(setCell(0, 3), applied).col).toBe(5);
  });

  it('mergeCells rectangle shifts like clearRange', () => {
    const merge: Op = { type: 'mergeCells', sheet: 's1', baseRev: 0, row: 2, col: 1, endRow: 4, endCol: 2 };
    const out = transform(merge, { type: 'insertRows', sheet: 's1', baseRev: 0, index: 3, count: 2 });
    expect(out.row).toBe(2);
    expect(out.endRow).toBe(6);
    // delete of the whole range collapses to a degenerate rectangle (applyOp no-ops it)
    const gone = transform(merge, { type: 'deleteRows', sheet: 's1', baseRev: 0, index: 2, count: 3 });
    expect(gone.row).toBe(2);
    expect(gone.endRow).toBe(2);
  });
});

File: lib/sheet/transform_test.go
package sheet

import "testing"

func TestTransformCellAgainstInsertRows(t *testing.T) {
	applied := Op{Type: OpInsertRows, Sheet: "s1", Index: 2, Count: 3}
	in := Op{Type: OpSetCell, Sheet: "s1", Row: 4, Col: 0, Raw: ptr("x")}
	out := Transform(in, applied)
	if out.Row != 7 {
		t.Fatalf("cell below insert must shift down by 3: got row %d", out.Row)
	}
	above := Transform(Op{Type: OpSetCell, Sheet: "s1", Row: 1, Col: 0, Raw: ptr("y")}, applied)
	if above.Row != 1 {
		t.Fatalf("cell above insert must not move: got row %d", above.Row)
	}
}

func TestTransformCellAgainstDeleteRows(t *testing.T) {
	applied := Op{Type: OpDeleteRows, Sheet: "s1", Index: 2, Count: 2} // deletes rows 2,3
	below := Transform(Op{Type: OpSetCell, Sheet: "s1", Row: 5, Col: 0, Raw: ptr("x")}, applied)
	if below.Row != 3 {
		t.Fatalf("cell below delete must shift up by 2: got %d", below.Row)
	}
	// A cell inside the deleted band: clamp to the deletion index (its row is gone).
	inside := Transform(Op{Type: OpSetCell, Sheet: "s1", Row: 3, Col: 0, Raw: ptr("z")}, applied)
	if inside.Row != 2 {
		t.Fatalf("cell inside deleted band should clamp to index 2: got %d", inside.Row)
	}
}

func TestTransformDifferentSheetIsNoop(t *testing.T) {
	applied := Op{Type: OpInsertRows, Sheet: "other", Index: 0, Count: 5}
	in := Op{Type: OpSetCell, Sheet: "s1", Row: 1, Col: 1, Raw: ptr("x")}
	out := Transform(in, applied)
	if out.Row != 1 || out.Col != 1 {
		t.Fatal("ops on different sheets must not transform")
	}
}

func TestTransformInsertAgainstInsert(t *testing.T) {
	applied := Op{Type: OpInsertRows, Sheet: "s1", Index: 2, Count: 2}
	in := Op{Type: OpInsertRows, Sheet: "s1", Index: 4, Count: 1}
	out := Transform(in, applied)
	if out.Index != 6 {
		t.Fatalf("later insert index must shift by applied count: got %d", out.Index)
	}
}

func TestTransformColsAgainstInsertCols(t *testing.T) {
	applied := Op{Type: OpInsertCols, Sheet: "s1", Index: 1, Count: 2}
	in := Op{Type: OpSetCell, Sheet: "s1", Row: 0, Col: 3, Raw: ptr("x")}
	out := Transform(in, applied)
	if out.Col != 5 {
		t.Fatalf("cell right of col insert must shift by 2: got %d", out.Col)
	}
}
